cluster/engine.go

```diff
@@ -417,10 +420,20 @@ func (e *Engine) CheckConnectionErr(err error) {
 	if err == nil {
 		e.setErrMsg("")
 		// If current state is unhealthy, change it to healthy
+		// A lock is needed to avoid race condition, which can lead to block on resumeCh
+		// "changed" flag will take care of slow operations, ex: emitEvent
+		changed := false
+		e.Lock()
```

Comment:
Why are you locking here?

Reply:
I thought of it, but the error message was a key to figure out the issue and saw the race condition possibility here. So figured why not fixing it as well as making sure only one refresh perform when the engine health changes.
What's the concerns with the lock?

Reply:
If anything else is holding the lock, `CheckConnectionErr` will get blocked.

Reply:
If this is an issue then we have a bigger problem to deal with , as subcommands in the same block also holds the lock, ex: `getFailureCount` and `resetFailureCount`

Reply:
Fair enough.